src/delay_history.jl

```
@@ -0,0 +1,168 @@
+        throw(AssertionError("Can only push times after the buffer end"))
+    end
+    push!(b.buff[indx], TimedVal(t, val))
+end
+
+# Get last value before time t
+function (b::DDEBuffer)(t, indx::Int)
+    bi = b.buff[indx]
+    v0 = nothing
+    t0 = nothing
+    v1 = nothing
+    t1 = nothing
+    v2 = nothing
+    t2 = nothing
+
+    if first(bi).t > t
+        throw(AssertionError("time $t is before start of buffer"))
+    end
+
+    for vt in bi
+        if vt.t <= t
+            # Keep updating as long as vt.t < t
+            # Save prev value
+            v0 = v1
+            t0 = t1
+            # Current value
+            v1 = vt.val
+            t1 = vt.t
+        else
+            # We found te last value, now get one more
+            v2 = vt.val
+            t2 = vt.t
+            break
+
+        end
+    end
+    # Error if no value found
+    v1 === nothing && throw(AssertionError("Value not in buffer range at t: $t"))
+
+    if t1 == t
+        # If exact value
+        return v1
+    elseif v2 === nothing
+        # If extrapolating forward
+        @warn "Interpolation inaccurate at t=$t"
+        return v1
+    elseif t2 <= t1
+        @warn "Interpolation found two equal time steps at t=$t"
+        return v1
+    elseif t2 in b.disc
+        # Next time point is a discontinuity
+        if v0 === nothing
+            # Not no provious point exists
+            @warn "Discontinuity found at next time point, no interpolation t=$t, t2=$t2"
+            return v1
+        elseif t0 == t1
+            # Previous point is same as current
+            @warn "Discontinuity found at next time point, two equal vals t=$t, t2=$t2"
+            return v1
+        else
+            # Extrapolate linearly from two previous points
+            return (v0*(t1 - t) + v1*(t - t0))/(t1-t0)
+        end
+    else
+        # Interpolate between current and next
+        return (v1*(t2 - t) + v2*(t - t1))/(t2-t1)
+    end
+    return v1 # Can't happen
```

Isn't it better to throw an error if this cant happen? Seems like a weird programming style, or is it  to avoid type instability?